pkg/command/apitoken.go

```diff
@@ -0,0 +1,89 @@
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/fatih/color"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+
+	"github.com/vmware-tanzu/tanzu-plugin-runtime/config"
+	"github.com/vmware-tanzu/tanzu-plugin-runtime/config/types"
+	"github.com/vmware-tanzu/tanzu-plugin-runtime/plugin"
+
+	commonauth "github.com/vmware-tanzu/tanzu-cli/pkg/auth/common"
+	"github.com/vmware-tanzu/tanzu-cli/pkg/auth/uaa"
+	"github.com/vmware-tanzu/tanzu-cli/pkg/cli"
+	"github.com/vmware-tanzu/tanzu-cli/pkg/constants"
+	timeutils "github.com/vmware-tanzu/tanzu-cli/pkg/utils/time"
+)
+
+func newAPITokenCmd() *cobra.Command {
+	apiTokenCmd := &cobra.Command{
+		Use:     "api-token",
+		Short:   "API Token Generation for the Tanzu Platform",
+		Aliases: []string{"apitoken"},
+		Annotations: map[string]string{
+			"group": string(plugin.SystemCmdGroup),
+		},
+	}
+
+	apiTokenCmd.SetUsageFunc(cli.SubCmdUsageFunc)
+	apiTokenCmd.AddCommand(
+		newAPITokenCreateCmd(),
+	)
+
+	return apiTokenCmd
+}
+
+func newAPITokenCreateCmd() *cobra.Command {
+	createCmd := &cobra.Command{
+		Use:     "create",
+		Short:   "Create new API Token",
+		Aliases: []string{},
+		RunE:    createAPIToken,
+	}
+
+	return createCmd
+}
+
+func createAPIToken(cmd *cobra.Command, _ []string) (err error) {
+	c, err := config.GetActiveContext(types.ContextTypeTanzu)
+	if err != nil {
+		return errors.New("No active context of type `tanzu`. Please login to Tanzu Platform first to generate the API token")
+	}
+	if c == nil || c.GlobalOpts == nil || c.GlobalOpts.Auth.Issuer == "" {
+		return errors.New("Invalid active context of type `tanzu`. Please login to Tanzu Platform first to generate the API token")
+	}
+	// Make sure it is of type tanzu with tanzuIdpType as `uaa` else return error
+	if idpType, exist := c.AdditionalMetadata[config.TanzuIdpTypeKey]; !exist || idpType != "uaa" {
+		return errors.New("invalid IDP type for the active context. Only UAA IDP type is supported for generating API token")
+	}
+
+	var token *commonauth.Token
+	// If user chooses to use a specific local listener port, use it
+	// Also specify the client ID to use for token generation
+	loginOptions := []commonauth.LoginOption{
+		commonauth.WithListenerPortFromEnv(constants.TanzuCLIOAuthLocalListenerPort),
+		commonauth.WithClientID(uaa.TanzuCLIClientIDExtended),
+	}
+
+	token, err = uaa.TanzuLogin(c.GlobalOpts.Auth.Issuer, loginOptions...)
+	if err != nil {
+		return errors.Wrap(err, "unable to login")
+	}
+	cyanBold := color.New(color.FgCyan).Add(color.Bold)
+	bold := color.New(color.Bold)
+
+	fmt.Fprint(cmd.OutOrStdout(), bold.Sprint("==\n\n"))
+	fmt.Fprintf(cmd.OutOrStdout(), "%s Your generated API token is: %s\n\n", bold.Sprint("API Token Generation Successful!"), cyanBold.Sprint(token.RefreshToken))
+	fmt.Fprintf(cmd.OutOrStdout(), "%s The token will expire on %s.\n", bold.Sprint("Token Expiration:"), bold.Sprint(timeutils.Now().Local().Add(time.Second*time.Duration(token.ExpiresIn)).Format(("2006-01-02 15:04:05"))))
```

Comment:
This is not quite right, unfortunately ExpiresIn only refers to validity of the access token. I suggest we remove this line,
then L86 should be next, followed by L85, which needs to be edited to remove reference to the expiration time.

Reply:
Do we know the expiration time easily?
Since we tell the user `regenerate a new token before expiration`, I find it nice to tell them when that will be.  If not, is there another way for them to avoid their usage failing when expiration happens?

Reply:
Apologies for the confusion and not using this thread to discuss this topic. I responded separately https://github.com/vmware-tanzu/tanzu-cli/pull/820#issuecomment-2397636551

Also, I have updated the PR description to remove the outdated screenshot and added the sample output (without color).